### ISSUE TITLE
`yapf: enable` comments inside expression cause infinite hang

### ISSUE THREAD

**Issue by danthedeckie** [psf/black#3026]
**Describe the bug**

If you have `yapf: enable` comments inside an expression, you can cause black to hang:

**To Reproduce**

with `file.py` as:
```python
def test_func():
    # yapf: disable
    if a:
        return True
    # yapf: enable
    elif b:
        return True

    return False
```

```sh
$ black file.py
```

It hangs forever.

**Expected behavior**

It should not hang.

At the very least it should give an error.

Preferably it should ignore those comments and not break anything.

**Environment**

- Black's version: black, 22.3.0 (compiled: yes)
- OS and Python version: Python 3.8 & 3.10, mac OS 12.3

Also happens on the https://black.vercel.app/

**Comment by danthedeckie** [comment]
```python
# yapf: disable
if a:
    return True
# yapf: enable
elif b:
    return False
```
is the smallest I can reduce it to.

It seems that the `yapf: disable` is needed as well.

**Comment by danthedeckie** [comment]
And version hunting, it seems (for me) to crop up between:
`black==19.10b0` (good - doesn't hang)
and
`black==20.8b0` (bad - does hang)

Hopefully that helps narrow it down.

**Comment by danthedeckie** [comment]
And trying to find the problem w/ `git bisect` narrows it down to:

```
892eddacd215d685e136686b7f629ade70adca83 is the first bad commit
commit 892eddacd215d685e136686b7f629ade70adca83
Author: otstrel <otstrel@gmail.com>
Date:   Fri May 8 15:37:17 2020 +0300

    Fix for "# fmt: on" with decorators (#1325)

 black.py                | 49 ++++++++++++++++++++++++++++++++++++++++---------
 tests/data/fmtonoff4.py | 31 +++++++++++++++++++++++++++++++
 tests/test_black.py     |  8 ++++++++
 3 files changed, 79 insertions(+), 9 deletions(-)
 create mode 100644 tests/data/fmtonoff4.py
```
